storage: heal shards that contain backwards incompatible errors

petrosagg opened the pull request:
### Motivation

This PR implements the remediation of having persisted backward incompatible protobuf data in certain errored upsert sources. I will explain the general strategy. The protobuf message that contained the backward incompatible field was `ProtoUpsertError` and specifically its `value` field that changed from being a `ProtoDataflowError` in `0.68.1` to being a `ProtoDecodeError` in `0.69.1`.

This PR is changing the `value` field to be of type `mz_storage_types.shim.ProtoUpsertValueErrorShim`. This type lives in its own file (`shim.proto`) which is necessary for `prost` to correctly generate the code but the generated `shim.rs` file is *not* included in the module tree with the `include!()` macro that we do normally. This allows us to autogenerate all the protobuf types with one hole that we need to fill in.

This PR includes a manually created `shim.rs` file that manually declares the `ProtoUpsertValueErrorShim` type to be an enum that has two variants, either `New` or `Legacy`. The `New` variant contains the real proto type that `0.69.1` uses and the `Legacy` variant contains the real proto type that `0.68.1` used (renamed to `ProtoUpsertValueErrorLegacy` in this PR).

This manual declaration also includes a manual `prost::Message` implementation which is what `prost` autogenerates for you from all the proto files. This `Message` implementation mostly delegates to the inner proto types by matching on the variant and forwarding the call. There is only one call that we override and that is `merge_field`.

`Message::merge_field` is called a protobuf message is being deserialized and its job is to mutate `self` and add information for the field in question. The `value` field of `ProtoUpsertError` (careful, this is NOT the same as `ProtoUpsertValueError`) has the tag number `1` for both new and legacy variants. Therefore, when `merge_fields` is called with `tag == 1` it's when we deploy a bunch of heuristics and application specific knowledge to understand if the bytes presented to us are of the new or the old variant.

If you're following along we've now have created a special `Proto` struct that can understand what type the bytes are presented to it and neatly encodes this information as an enum.

Moving along, we need to look at `UpsertvalueError` the "real" rust struct that we work with in memory. This struct gained a new field called `is_legacy_dont_touch_it` which is set by its `RustType::from_proto` implementation. This is simply turning the `enum` information that existed in `ProtoUpsertValueErrorShim` into a boolean field. The reason for not making `UpsertValueError` also an enum is due to the number of code changes that would have to go with it.

Ok, so now we have an in-memory `UpsertValueError` type that can remember if it was created by a legacy encoding. The final piece of the recipe is in the upsert operator. When the upsert operator starts up and rehydrates itself from a snapshot it takes note of all the legacy errors that it got and their associated diffs. When the snapshot phase is over, the operator will emit retractions (i.e negated diffs) for all observed legacy errors and insertions for non-legacy ones. This means that shortly after that moment we will permanently record in persist a retraction of the old incompatible bytes and an insertion of the new, correct bytes.

Finally, when compaction eventually catches up and goes over these diffs it will completely eliminate the legacy encoding and at that point we should be able to delete the monstrosity presented here.



<!--
Which of the following best describes the motivation behind this PR?

  * This PR fixes a recognized bug.

    [Ensure issue is linked somewhere.]

  * This PR adds a known-desirable feature.

    [Ensure issue is linked somewhere.]

  * This PR fixes a previously unreported bug.

    [Describe the bug in detail, as if you were filing a bug report.]

  * This PR adds a feature that has not yet been specified.

    [Write a brief specification for the feature, including justification
     for its inclusion in Materialize, as if you were writing the original
     feature specification.]

   * This PR refactors existing code.

    [Describe what was wrong with the existing code, if it is not obvious.]
-->

### Tips for reviewer

The first commit is a revert of the original fix deployed to production so only the diffs of the second commit should be reviewed

<!--
Leave some tips for your reviewer, like:

    * The diff is much smaller if viewed with whitespace hidden.
    * [Some function/module/file] deserves extra attention.
    * [Some function/module/file] is pure code movement and only needs a skim.

Delete this section if no tips.
-->

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

antiguru commented on src/storage-types/src/errors.rs:
```diff
@@ -429,6 +468,8 @@ mod columnation {
         /// Region for storing u8 vectors.
         u8_region: <Vec<u8> as Columnation>::InnerRegion,
+        /// Region for storing bools.
+        bool_region: <bool as Columnation>::InnerRegion,
```
You don't need to store `bool` in a region since it won't contain any heap-allocated data!

petrosagg replied:
Ah, yes. I got confused by the existence of the `u8` region but that is `Vec<u8>` not just `u8`. Removed!

teskje commented on src/storage-types/src/shim.rs:
```diff
@@ -0,0 +1,543 @@
+            // make a local copy. This would normally be tricky with only the `Buf` bound but we
+            // know that we only get here as a result of a `Codec::decode` call, which has a &[u8]
+            // argument that is what B will be. This means that the `B: Buf` generic type is backed
+            // by a contiguous buffer and therefore calling `Buf::chunk` will give us all the data.
```
> we only get here as a result of a `Codec::decode` call

That's not entirely true, right? `SourceDataDecode` calls `ProtoDataflowError::decode` directly. It passes a `&[u8]` so the rest of the reasoning still holds, but I'm a bit worried that we are overlooking over calls to `Message::decode` on one of the types that contain a shim. Is there any way we can assert that we really get all the data here?

teskje replied:
Alternatively, we could follow the simpler strategy of trying the decoding and `is_some` check for the new variant first, and if that fails decode with the old variant. AFAICT the only difference to the current behavior would be that things that successfully decode under both variants would not produce a panic but would be encoded with the new variant instead. Which seems fine?

petrosagg replied:
>  Is there any way we can assert that we really get all the data here?

Good idea, I added an assert that checks `data.len()` against `buf.remaining()`.

> Alternatively, we could follow the simpler strategy of trying the decoding and is_some check for the new variant first, and if that fails decode with the old variant

This isn't possible to do in here because the `buf` argument only implements the `Buf` trait, and you cannot use that trait without consuming its data. Were you thinking of something else?

teskje replied:
> Were you thinking of something else?

Nope, I just wasn't aware of that limitation.

teskje reviewed (commented):
I think I can convince myself that this approach should work! But I'm not a protobuf expert, so I can't review the `Message` impl with high confidence.

def- commented on src/storage-types/src/errors.rs:
```diff
@@ -163,40 +165,77 @@ pub struct UpsertValueError {
+impl RustType<ProtoUpsertValueErrorShim> for UpsertValueError {
+    fn into_proto(&self) -> ProtoUpsertValueErrorShim {
+        if self.is_legacy_dont_touch_it {
+            let inner = ProtoDataflowError {
+                kind: Some(proto_dataflow_error::Kind::DecodeError(
+                    self.inner.into_proto(),
+                )),
+            };
+            let proto = ProtoUpsertValueErrorLegacy {
+                inner: Some(inner),
+                for_key: Some(self.for_key.into_proto()),
+            };
+            ProtoUpsertValueErrorShim::Legacy(Box::new(proto))
```
Uncovered

def- commented on src/storage-types/src/errors.rs:
```diff
@@ -163,40 +165,77 @@ pub struct UpsertValueError {
+                    .into_rust_if_some("ProtoUpsertValueError::for_key")?,
+                is_legacy_dont_touch_it: false,
+            },
+            ProtoUpsertValueErrorShim::Legacy(legacy) => {
+                let legacy_inner = match legacy.inner {
+                    Some(inner) => inner,
+                    None => {
+                        return Err(TryFromProtoError::missing_field(
+                            "ProtoUpsertValueError::inner",
+                        ))
+                    }
+                };
+                let inner = DataflowError::from_proto(legacy_inner)?;
+                let inner = match inner {
+                    DataflowError::DecodeError(inner) => *inner,
+                    _ => panic!("invalid legacy upsert error"),
+                };
+                UpsertValueError {
+                    inner,
+                    for_key: legacy
+                        .for_key
+                        .into_rust_if_some("ProtoUpsertValueError::for_key")?,
+                    is_legacy_dont_touch_it: true,
```
Uncovered

def- commented on src/storage/src/render/upsert.rs:
```diff
@@ -487,6 +488,20 @@ where
+                    // that time we will have emitted the error correction, which happens before
+                    // processing any of the new source input.
+                    if err.is_legacy_dont_touch_it {
+                        legacy_errors_to_correct.push((err.clone(), diff.clone()));
+                        err.is_legacy_dont_touch_it = false;
```
Uncovered

def- commented on src/storage/src/render/upsert.rs:
```diff
@@ -531,6 +546,19 @@ where
+        // them at because all they do is change the representation. The error count at any
+        // timestamp remains constant.
+        for (mut error, diff) in legacy_errors_to_correct {
+            assert!(error.is_legacy_dont_touch_it, "attempted to correct non-legacy error");
+            let time = output_cap.time().clone();
+            let retraction = Err(UpsertError::Value(error.clone()));
+            error.is_legacy_dont_touch_it = false;
+            let insertion = Err(UpsertError::Value(error));
+            output_handle.give(&output_cap, (retraction, time.clone(), -diff)).await;
+            output_handle.give(&output_cap, (insertion, time, diff)).await;
```
Uncovered

def- reviewed (changes requested):
Possible to write an explicit test for this by having an old-new proto file around and checking the handling?

Also marked some spots inline based on coverage report: https://buildkite.com/materialize/coverage/builds/231

Nightly failures look unrelated to this change: https://buildkite.com/materialize/nightlies/builds/4237

petrosagg commented:
> Possible to write an explicit test for this by having an old-new proto file around and checking the handling?

The PR includes a unit test that does this with many different kinds of structs. See the bottom of `shim.rs`. Were you thinking something more than that?

def- commented:
I was thinking an end-to-end test, but unit test is better than nothing. Are the coverage holes expected with this unit test?

guswynn commented on src/storage-types/src/shim.proto:
```diff
@@ -0,0 +1,20 @@
+
+package mz_storage_types.shim;
+
+// This is just a shim message to try and get prost to generate a reference to
+// this shim as part of the `prost::Message` implementation for the
+// ProtoUpsertError type.
+//
+// The actual definition of `ProtoUpsertValueErrorShim` and its implementation
+// of `prost::Message` happens manually in `shim.rs`
+message ProtoUpsertValueErrorShim { }
```
You need to document what `prost` command was used to generate the original version of `shim.rs`

petrosagg replied:
Hm, `shim.rs` was not created based on some original `prost` generated file. I just wrote a new rust module containing the `ProtoUpsertValueErrorShim` type and implemented the traits

guswynn commented on src/storage/src/render/upsert.rs:
```diff
@@ -487,6 +488,20 @@ where
+                    // that time we will have emitted the error correction, which happens before
+                    // processing any of the new source input.
+                    if err.is_legacy_dont_touch_it {
+                        legacy_errors_to_correct.push((err.clone(), diff.clone()));
```
technically this is an unbounded buffer, so a pathological spill2disk instance could oom...probably not worth working around

Also, is the idea that retracted errors will have 2 opposite retractions sent below?

petrosagg replied:
I need to add a consolidation step below but the idea is that with whatever diff a legacy errors comes in with we must emit it with that diff negated. So yes, if we see `(legacy_error, -1)` we will emit `(legacy_error, +1, new_error, -1)`. One way this can happen is if we successfully submit the corrections, then we crash, then on restart we read a snapshot that includes the original legacy error AND our legacy error correction. So we must not assume anything about diffs

guswynn reviewed (commented):
I think the heuristics are fine, and this is a reasonable approach, but as far as I can tell, we don't have any metrics on if a shard has old error values inside it. Even if we added some in upsert rehydration, sources that are not currently ingesting (if there is no active replica) wouldn't be covered (we could probably find all those though)

philip-stoev commented on src/testdrive/Cargo.toml:
```diff
@@ -1,7 +1,7 @@
 name = "mz-testdrive"
 description = "Integration test driver for Materialize."
-version = "0.0.0"
+version = "0.70.0-dev"
```
If this is going to be versioned going forward, please add this to `bin/bump-version` , line 32

philip-stoev replied:
Oh you did that already, thanks!

philip-stoev commented on test/incident-72/mzcompose.py:
```diff
@@ -0,0 +1,57 @@
+    c.run("testdrive", "verify-heal-in-memory.td")
+    # c.kill("materialized")
+
+    # # Now run a version of Materialize that panics if it ever encounters a
```
This is commented out. Also, consider adding the workflow to Nightly so that it actually runs as a regression test.

petrosagg commented:
This is now ready to go!

@def- @philip-stoev I have added the incident-72 mzcompose test in the nightly test suite. This will exercise all the lines marked as uncovered by the previous coverage test.

@guswynn I have added a metric so that we know when it's safe to revert this

petrosagg commented:
@danhhz I changed the `force_compaction` function to be generic over the keys and values so that I can call it correctly from testdrive. `persist-cli` keeps providing the same types that were previously hardcoded. If you could take a look at the persist changes it would be great!

danhhz reviewed (commented):
src/persist* changes lgtm

guswynn commented:
metrics look good!
